File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/Project.java
package net.nemerosa.ontrack.model.structure;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.experimental.Wither;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.YesNo;

@Data
@AllArgsConstructor(access = AccessLevel.PUBLIC)
public class Project implements ProjectEntity {

    private final ID id;
    private final String name;
    private final String description;
    @Wither
    private final boolean disabled;
    @Wither
    private final Signature signature;

    public Project withId(ID id) {
        return new Project(id, name, description, disabled, signature);
    }

    public static Project of(NameDescriptionState nameDescription) {
        return new Project(ID.NONE, nameDescription.getName(), nameDescription.getDescription(), nameDescription.isDisabled(),
                Signature.none());
    }

    public static Project of(NameDescription nameDescription) {
        return new Project(ID.NONE, nameDescription.getName(), nameDescription.getDescription(), false,
                Signature.none());
    }

    @Override
    public Project getProject() {
        return this;
    }

    @Override
    public ProjectEntityType getProjectEntityType() {
        return ProjectEntityType.PROJECT;
    }

    @Override
    public String getEntityDisplayName() {
        return String.format("Project %s", name);
    }

    public static Form form() {
        return Form.nameAndDescription()
                .with(
                        YesNo.of("disabled").label("Disabled").help("Check if the project must be disabled.")
                );
    }

    public Project update(NameDescriptionState form) {
        return of(form).withId(id).withDisabled(form.isDisabled());
    }

    public Form asForm() {
        return form().name(name).description(description).fill("disabled", disabled);
    }
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/structure/Project.kt
package net.nemerosa.ontrack.model.structure

import com.fasterxml.jackson.annotation.JsonProperty

/**
 * A project in Ontrack. Usually associated with a SCM repository, this is the main entity in the Ontrack
 * model.
 */
data class Project(
        override val id: ID,
        val name: String,
        override val description: String?,
        @JsonProperty("disabled")
        val isDisabled: Boolean,
        override val signature: Signature
) : ProjectEntity {

    fun withDisabled(isDisabled: Boolean) = Project(id, name, description, isDisabled, signature)

    fun withSignature(signature: Signature) = Project(id, name, description, isDisabled, signature)

    fun withDescription(description: String) = Project(id, name, description, isDisabled, signature)

    fun withId(id: ID) = Project(id, name, description, isDisabled, signature)

    companion object {

        /**
         * Maximum length for the name of a project
         */
        const val PROJECT_NAME_MAX_LENGTH = 80

        @JvmStatic
        fun of(nameDescription: NameDescriptionState) =
                Project(ID.NONE, nameDescription.name, nameDescription.description, nameDescription.isDisabled, Signature.anonymous())

        @JvmStatic
        fun of(nameDescription: NameDescription) =
                Project(ID.NONE, nameDescription.name, nameDescription.description, false, Signature.anonymous())

    }

    override val project: Project get() = this

    override val entityDisplayName: String get() = "Project $name"

    override val parent: ProjectEntity? = null

    override val projectEntityType: ProjectEntityType = ProjectEntityType.PROJECT


    fun update(form: NameDescriptionState): Project =
            of(form).withId(id).withDisabled(form.isDisabled)

}
